### PR TITLE
initial implementation of appstore webhooks

### PR DESCRIPTION
**Pull request by husobee**
### Summary

adds appstore webhooks for skus

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

**Review by clD11 — CHANGES_REQUESTED** [review]

File: services/skus/controllers.go [hunk 1]
```diff
@@ -871,6 +871,7 @@ func WebhookRouter(service *Service) chi.Router {
 	r := chi.NewRouter()
 	r.Method("POST", "/stripe", middleware.InstrumentHandler("HandleStripeWebhook", HandleStripeWebhook(service)))
 	r.Method("POST", "/android", middleware.InstrumentHandler("HandleAndroidWebhook", HandleAndroidWebhook(service)))
+	r.Method("POST", "/ios", middleware.InstrumentHandler("HandleIOSWebhook", HandleIOSWebhook(service)))
```
Comment (clD11) [inline services/skus/controllers.go:874]: Maybe we could add a test for the handler

**Review by clD11 — CHANGES_REQUESTED** [review]

File: services/skus/controllers.go [hunk 2]
```diff
@@ -922,6 +923,61 @@ func HandleAndroidWebhook(service *Service) handlers.AppHandler {
...
+			validationErrMap["invalid-transaction-info"] = err.Error()
+		}
+
+		// if we had any validation errors, return the validation error map to the caller
+		if len(validationErrMap) != 0 {
+			return handlers.ValidationError("Error validating request url", validationErrMap)
+		}
+
+		if err := service.verifyIOSNotification(ctx, txInfo, renewalInfo); err != nil {
+			return handlers.WrapError(err, "failed to verify ios notification", http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/controllers.go:975]: Could also log the error

**Review by clD11 — CHANGES_REQUESTED** [review]

File: services/skus/service.go [hunk 3]
```diff
@@ -1174,6 +1176,37 @@ func (s *Service) verifyCredential(ctx context.Context, req credential, w http.R
...
 }
 
+// verifyIOSNotification - verify the developer notification from appstore
+func (s *Service) verifyIOSNotification(ctx context.Context, txInfo *appstore.JWSTransactionDecodedPayload, renewalInfo *appstore.JWSRenewalInfoDecodedPayload) error {
+	if txInfo == nil || renewalInfo == nil {
+		return errors.New("notification has no tx or renewal")
+	}
+
+	// lookup the order based on the token as externalID
+	o, err := s.Datastore.GetOrderByExternalID(txInfo.OriginalTransactionId)
```
Comment (stoletheminerals) [inline services/skus/service.go:1186]: It seems that `GetOrderByExternalID` is using parametrised queries. Just wanted to double check if we need to sanitise `OriginalTransactionId` here

**Review by stoletheminerals — COMMENTED** [review]
Left few inline comments/questions. Few additional questions I have:
1. Can we limit IP addresses that can access `/ios` endpoint?
2. Do we have a sample of HTTP request from Apple? I couldn't find it on internet.
3. I would add a test with a self-signed certificate to check that the endpoint would return an error during certificate validation.

File: services/skus/input.go [hunk 4]
```diff
@@ -300,3 +302,173 @@ func (an *AndroidNotification) Validate(ctx context.Context) error {
...
+		logger = logging.Logger(ctx, "IOSNotification.GetTransactionInfo")
+	)
+
+	// get the cert from jws header
+	rootCertStr, err := extractHeaderByIndex(iosn.SignedPayload, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	intermediaCertStr, err := extractHeaderByIndex(iosn.SignedPayload, 1)
```
[the remark this answers is not in the document]
Comment (husobee) [reply services/skus/input.go:425]: https://developer.apple.com/documentation/appstoreservernotifications/jwsdecodedheader

Based on the apple documentation,  the header at most has three items, root cert, intermediate certificate, cert with public key.  I believe the intermediate cert chain is included in the intermediate cert presented in the header as defined in the above link.

File: services/skus/receipt.go [hunk 5]
```diff
@@ -117,3 +181,103 @@ func validateAndroidReceipt(ctx context.Context, receipt interface{}) (string, e
...
+	_, err = cert.Verify(opts)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// rootPEM is from `openssl x509 -inform der -in AppleRootCA-G3.cer -out apple_root.pem`
+const appleRootPEM = `
```
[the remark this answers is not in the document]
Comment (stoletheminerals) [reply services/skus/receipt.go:267]: Not much of an issue from a security perspective. Just curious how we would know if we need to change this certificate in future. Root CAs usually have a long lifetime anyway, so we probably don't need to worry too much about it.

File: services/skus/controllers.go [hunk 6]
```diff
@@ -956,14 +956,14 @@ func HandleIOSWebhook(service *Service) handlers.AppHandler {
 		// transaction info
 		txInfo, err := req.GetTransactionInfo(ctx)
 		if err != nil {
-			logger.Warn().Err(err).Msg("failed to get renewal info from message")
+			logger.Warn().Err(err).Msg("failed to get transaction info from message")
 			validationErrMap["invalid-renewal-info"] = err.Error()
```
Comment (stoletheminerals) [inline services/skus/controllers.go:960]: `validationErrMap` is mismatched as well

Comment (husobee) [reply services/skus/controllers.go]: resolved: https://github.com/brave-intl/bat-go/pull/1649/commits/9064cd764800c268686b31b2d3153500162fca31

**Review by clD11 — DISMISSED** [review]

File: services/skus/service.go (hunk 3 above)
Comment (husobee) [reply services/skus/service.go:1196]: added validation of that originial transaction id in this commit: https://github.com/brave-intl/bat-go/pull/1649/commits/e54c7768b5be2af0b57995d0c9c8424fb4b41158

**Review by clD11 — APPROVED** [review]

**Review by stoletheminerals — APPROVED** [review]

**Pull request merged**